### ISSUE TITLE
A test function dying causes the loss of previous results

### ISSUE THREAD

**Issue by hmijail** [ThrowTheSwitch/Ceedling#72]
```
...
Linking test_TBuffer.out...
Running test_TBuffer.out...
sh: line 1: 58227 Segmentation fault: 11  ../../../build/test/out/test_TBuffer.out 2>&1

ERROR: Test executable "test_TBuffer.out" failed.
> Produced no output to $stdout.
> And exited with status: [139] (count of failed tests).
> This is often a symptom of a bad memory access in source or test code.

rake aborted!

/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator_helper.rb:36:in `test_results_error_handler'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator.rb:153:in `generate_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/rules_tests.rake:42:in `block in <top (required)>'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/task_invoker.rb:74:in `invoke_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:68:in `block in setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:32:in `setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/tasks_tests.rake:8:in `block (2 levels) in <top (required)>'
Tasks: TOP => ../../../build/test/results/test_TBuffer.pass
(See full trace by running task with --trace)

--------------------
OVERALL TEST SUMMARY
--------------------

No tests executed.
```

At the point of failure, 5 tests had already run (I know because the test function failing is the last one I added). No results are gathered, and no hint appears on what test function has failed.
Furthermore, the Test Summary is wrong.

**Comment by mvandervoord** [comment]
This isn't really the place for us to help you with debugging your tests. The forum is much better for that.

That being said, most often, this is the result of misusing pointers in your test, resulting in your test segfaulting. Sometimes people pass an address to a pointer when they meant to pass the pointer itself or vice versa... these are easy things to let slip by when you're working on a test, so they're usually the first guess I have as to what went wrong.

If that doesn't seem to apply, I'd recommending posting a snippet of the test on the forum with the output above. Myself or someone else will likely reply soon to try to help you out.

**Comment by hmijail** [comment]
Sorry, it's not that I'm asking for help debugging; I'm reporting a bug: Ceedling should report that a test function has died (which surely means a failed test!).

For context, the cause of the test function dying is that it passed a NULL pointer to the function being tested. So this is an example of a valid test and a valid, valuable failure. The problem is that the test failure is killing Ceedling. Instead, Ceedling should be able to report the failure.

**Comment by hmijail** [comment]
I have seen other C unit testing frameworks which use fork() to isolate the framework from a test dying like this. I guess that a more straightforward fix for Ceedling would be to create test runners which each isolates one test function... 
... but even simpler and still useful would be to just report what was the last test that started running before the runner died.

**Comment by mvandervoord** [comment]
Ah. Ceedling differentiates between test failures and test errors. Since the executable actually crashed, it appropriately calls it a test error. Ceedling doesn't have enough context information to know if the test caused the error or the release code did. So it just tells you that it couldn't run the test at all because it crashed.

We've talked about having an option for Unity to actually catch runtime exceptions like this, but they are SO host and target dependent that there isn't a tidy way to do this. So we just have to leave it to the developer to sort through his own major coding errors.

**Comment by mvandervoord** [comment]
Yeah, we are purposefully targeting Embedded developers. Many embedded targets don't have fork as an option, so we can't really standardize on that.

Ceedling COULD actually scan the output as it runs to report which test was running, I believe. I agree that this would be very useful. :)

**Comment by 0ge** [comment]
I would actually think this was really useful. I occasionally (probably more often than I dare to admit) write code that breaks a previous test with this. Then I don't know if there is something wrong with the test and function I was just writing, or if I broke some previous test/code.

For instance, if I'm writing FunctionA and TestA, get them green and good, then go on to write FunctionB, which causes (e.g. incorrect pointer or index management) FunctionA to fail (even though FunctionB doesn't call FunctionA), I have no idea what actually went wrong. If Ceedling instead said that TestA crashed, I know where to start looking for my mistake.

**Comment by hmijail** [comment]
But note that Ceedling is neither reporting the tests that were run correctly until the runner died, nor that this is a failure: it just dies.

I can kind-of fix it myself by adding a simple printf("%s\n",__func__) at the beginning of each of my tests, so I can see what was last running. But then... am I not doing manually what Ceedling should be doing?

**Comment by mvandervoord** [comment]
;) You've both convinced me. I agree that it's worth having Ceedling spelunk through the output to give more meaningful context here.

**Comment by hmijail** [comment]
For the record: the printf I mentioned does not work because Ceedling reports "> Produced no output to $stdout." . Which is false!

**Comment by hmijail** [comment]
(That is, the printf IS working, but when the test runner dies, Ceedling does not show the collected stdout)

**Comment by hmijail** [comment]
Not wanting to be pushy, but... any idea about when this might be implemented? 

I need to know whether I should wait a bit (and so I can work around this problem for now) or a lot (and then maybe I should try to solve it myself and contribute a patch?)

**Comment by mvandervoord** [comment]
Sorry, I had assumed no one was actually waiting on this one since it seemed to just be a nicety feature. I had pushed it to the bottom of the priority list.

Wanna do an experiment? If you have a test that currently has this issue (or can reproduce it quickly), then in your Ceedling install in /lib/ceedling/tool_executor.rb around line 78, you should see a line that looks something like this:

raise ShellExecutionException.new(shell_result) if ((shell_result[:exit_code] != 0) and options[:boom])

Just comment it out by adding a # in front of it... then rerun your tests and see what happens. Does it do more or less what you want in that case? I'm trying to figure out a simple solution here. This isn't the solution, but it might provide a clue as to the direction we can take.

**Comment by hmijail** [comment]
> On 05 May 2016, at 20:32, Mark VanderVoord <notifications@github.com> wrote:
> 
> Sorry, I had assumed no one was actually waiting on this one since it seemed to just be a nicety feature. I had pushed it to the bottom of the priority list.
> 
> Wanna do an experiment? If you have a test that currently has this issue (or can reproduce it quickly),
> 
It's easy to reproduce:

void test_NULLptr(void){
    printf("%s\n",__func__);
    int *p = NULL;
    *p = 0;
}

... assuming that it's a platform that will not allow writing to 0, of course. But in general, if you use gcc or compatibles, you could just call __builtin_trap(), which kills the program.

> then in your Ceedling install in /lib/ceedling/tool_executor.rb around line 78, you should see a line that looks something like this:
> 
> raise ShellExecutionException.new(shell_result) if ((shell_result[:exit_code] != 0) and options[:boom])
> 
> Just comment it out by adding a # in front of it... then rerun your tests and see what happens. Does it do more or less what you want in that case? I'm trying to figure out a simple solution here. This isn't the solution, but it might provide a clue as to the direction we can take.
> 
I just tried that, and I don't see any difference :(.

Running test_TBuffer.out...
sh: line 1:  9845 Illegal instruction: 4  ../../../build/test/out/test_TBuffer.out 2>&1

ERROR: Test executable "test_TBuffer.out" failed.
> Produced no output to $stdout.
> And exited with status: [132] (count of failed tests).
> This is often a symptom of a bad memory access in source or test code.

rake aborted!

/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator_helper.rb:36:in `test_results_error_handler'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator.rb:153:in `generate_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/rules_tests.rake:42:in `block in <top (required)>'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/task_invoker.rb:74:in `invoke_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:68:in `block in setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:32:in `setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/tasks_tests.rake:8:in `block (2 levels) in <top (required)>'
Tasks: TOP => ../../../build/test/results/test_TBuffer.pass
(See full trace by running task with --trace)

--------------------
OVERALL TEST SUMMARY
--------------------

No tests executed.

**Comment by hmijail** [comment]
(reposting because the answer via mail got all broken)

It's easy to reproduce:

```
void test_NULLptr(void){
    printf("%s\n",__func__);
    int *p = NULL;
    *p = 0;
}
```

... assuming that it's a platform that will not allow writing to 0, of course. But in general, if you use gcc or compatibles, you could just call __builtin_trap(), which kills the program.

I just tried the edit you said, and I don't see any difference :(.

```
Running test_TBuffer.out...
sh: line 1:  9845 Illegal instruction: 4  ../../../build/test/out/test_TBuffer.out 2>&1

ERROR: Test executable "test_TBuffer.out" failed.
> Produced no output to $stdout.
> And exited with status: [132] (count of failed tests).
> This is often a symptom of a bad memory access in source or test code.

rake aborted!

/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator_helper.rb:36:in `test_results_error_handler'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator.rb:153:in `generate_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/rules_tests.rake:42:in `block in <top (required)>'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/task_invoker.rb:74:in `invoke_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:68:in `block in setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:32:in `setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/tasks_tests.rake:8:in `block (2 levels) in <top (required)>'
Tasks: TOP => ../../../build/test/results/test_TBuffer.pass
(See full trace by running task with --trace)

--------------------
OVERALL TEST SUMMARY
--------------------

No tests executed.
```

**Comment by hmijail** [comment]
An idea about how to fix it: I see that <ceedling gem>/lib/ceedling/generator_test_runner.rb generates the test runners (duh ;P), and in line 58 it does
```
    @test_runner_generator.generate( module_name,
                                     runner_filepath,
                                     test_cases,
                                     mock_list,
                                     test_file_includes)
```
If I understand correctly, Ceedling generates one test runner per test_*.c file. Why not generate instead one test runner per test function? The module_name could be changed to module_name+test_case[n], and the result parsing should be changed to only accept exit value 0 as success, and anything else as failure. This way, any test dying would rightly be caught as a failure, and next tests would continue running without any problem. Furthermore individual tests would be more reliable. I guess that the rest of Ceedling would need no changes.

And additionally issue #71 would be fixed as a side effect ;).

**Comment by Letme** [comment]
That is going to increase execution time on simulators (flashing/loading
each test-file), so it should be at least optional if nothing else.

**Comment by mvandervoord** [comment]
Ha, no. The runner generator's purpose is to create a runner for a collection of tests. If we did what you are suggesting you would get a test summary for every individual test, which isn't what we are looking for here. We're looking to just catch the broken executable and report its progress before reporting that it died. 

It's interesting that this last tweak didn't give you results. Maybe your platform isn't sending them to stdout until the program actually finishes. In this case it might be interesting to try redefining  UNITY_OUTPUT_CHAR(a) to be something like { putchar(a); fflush(stdout); } just to see if that works here. Again, this is more information gathering than it is a real solution. Many platforms couldn't do this.

**Comment by hmijail** [comment]
Tried defining UNITY_OUTPUT_CHAR(a) as you said: no changes. 

BUT I tried adding a `fflush(stdout);` as the first line of my test, and now the run of Ceedling shows all the text that before was not appearing. 

```
Running test_TBuffer.out...
sh: line 1: 10547 Illegal instruction: 4  ../../../build/test/out/test_TBuffer.out 2>&1

ERROR: Test executable "test_TBuffer.out" failed.
> Produced no final test result counts in $stdout:
test_TBuf_Create_OK
test_TBuffer.c:82:test_TBuf_Create_OK:PASS
test_TBuf_Create_failFirstMalloc
0/F/TBuffer.c:24/Failed to malloc TBuf: No Memory
test_TBuffer.c:94:test_TBuf_Create_failFirstMalloc:PASS
test_This_is_a_failing_test
test_TBuffer.c:105:test_This_is_a_failing_test:FAIL:Function tmalloc.  Called less times than expected.
test_TBuf_Create_failSecondMalloc
0/F/TBuffer.c:35/Failed to malloc TBuf's mutex: No Memory
test_TBuffer.c:112:test_TBuf_Create_failSecondMalloc:PASS
test_TBuf_Create_failThirdMalloc
0/F/TBuffer.c:49/Failed to malloc TBuf ref count: No Memory
test_TBuffer.c:125:test_TBuf_Create_failThirdMalloc:PASS
test__TBuf_Copy_nullINnullOUT
test_TBuffer.c:165:test__TBuf_Copy_nullINnullOUT:PASS
test_TBuf_Copy_All_singleTBuf
test_TBuffer.c:174:test_TBuf_Copy_All_singleTBuf:PASS
> And exited with status: [132] (count of failed tests).
> This is often a symptom of a bad memory access in source or test code.

rake aborted!

/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator_helper.rb:36:in `test_results_error_handler'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/generator.rb:153:in `generate_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/rules_tests.rake:42:in `block in <top (required)>'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/task_invoker.rb:74:in `invoke_test_results'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:68:in `block in setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/test_invoker.rb:32:in `setup_and_invoke'
/Users/mija/.gem/ruby/2.0.0/gems/ceedling-0.20.3/lib/ceedling/tasks_tests.rake:8:in `block (2 levels) in <top (required)>'
Tasks: TOP => ../../../build/test/results/test_TBuffer.pass
(See full trace by running task with --trace)

--------------------
OVERALL TEST SUMMARY
--------------------

No tests executed.
```

**Comment by hmijail** [comment]
I should be more precise: the fflush() in my test causes that output without having to define UNITY_OUTPUT_CHAR.
As for that define, I did it on top of my test_*.c file, just before the `#include "unity.h"`. I hope that was enough.

**Comment by mvandervoord** [comment]
:) No, just putting the definition for UNITY_OUTPUT_CHAR() into your test file doesn't do it. It needs to be passed to the compiler so that when unity.c is built, it gets used.

There's no need to do that, now, though. Your followup experiment got all the information I needed. Nice job!  :)

From your experiments, we've learned that the data isn't getting flushed as the test progresses. So, if the executable fails, the stdout stream is never getting out of the exe. So one step towards fixing this is making sure that the flush happens. Then, the second part of the fix is going to be sifting through the output even when an error code was returned so that we can return the previous results. Finally (step three), we are going to append the new error onto the last test that was started but never finished.

I'm hoping to tackle this tomorrow. Sorry about the delays. Thanks for your help!

**Comment by hmijail** [comment]
That would be awesome! Thanks!

And sorry about the #define mix-up, I assumed/hoped it would be picked up by unity.h.  (Maybe that could be documented too? When it's mentioned in http://www.throwtheswitch.org/build/cross, it rather looks like it might be enough to do as I did)

**Comment by mvandervoord** [comment]
Alrighty Implementing the first step in my description from yesterday is complete. This produces output now which has the normal results and then says:

```
And then likely crashed. 
This is often a symptom of a bad memory access in source or test code.
```

This appears like a good way to handle this. It's narrowed it down to a particular test and then given a hint as to what happened. So I'm going with that. See version commit 68322a8 :)

I've also fixed the side-note in the docs you mentioned above. Thanks for the help!

**Comment by hmijail** [comment]
Could this issue stay open until the 2nd and 3rd steps are implemented? That would make easier to follow any evolution. 
Or should I open independent issues for those steps?

**Comment by mvandervoord** [comment]
Steps 2 & 3 weren't necessary in the end. It not reports the last test to complete and tells you it crashes after that. This is what 2 & 3 were meant to do.

**Comment by Miv13** [comment]
Has this  functionality been reverted? As of now (v. 0.28.3) I only get:

`Running test_unit_flashdrv.out...

ERROR: Test execuERROR: Test executable "test_unit_flashdrv.out" failed.
> Produced no output to $stdout.
> And exited with status: [0] (count of failed tests).
> This is often a symptom of a bad memory access in source or test code.

rake aborted!
`

And there certainly were things to output.

**Comment by mvandervoord** [comment]
This functionality is still running. I just verified it still works on my setup.

The error you're getting is that no standard output was produced at all, instead of the cout being off. 

Is it the first test that is failing (so nothing has been output yet?

If you define `UNITY_OUTPUT_FLUSH` to `flush` your output, does that help?

**Comment by Miv13** [comment]
It's been a while, so I don't remember exactly, but judging by my own "And there certainly were things to output." I would guess that this was not the first test. I'll try to reproduce this and check the option that you mentioned. Do I just define it in project.yml?

**Comment by Miv13** [comment]
OK, I tested both cases with a bad memory access set after a specific number of tests.

In the default case, a bunch of results was outputed after (what I assume to be) filling a buffer - if I set the bad access in 31st test, I got no output, but if I set it in 61st test, I got the results for the first 30-40 tests (the number varying on the amount of 'prints' I had in the tests).

If I define `UNITY_USE_FLUSH_STDOUT` (which in turn sets `UNITY_OUTPUT_FLUSH()` to `(void)fflush(stdout)`), I get the output from all tests prior to the 'corrupted' one (but still without the fancy layout etc.).

Thanks for pointing me in the right direction, I guess this is good enough for me.
